asustor_it87.c

```diff
@@ -2641,6 +2651,277 @@ static SENSOR_DEVICE_ATTR(in8_label, S_IRUGO, show_label, NULL, 2);
+		   which apparently means "always on", regardless of bits 3:1 and 7:6
+		   (so leave them at 0) */
+		return BIT(0);
+	if(mode < 8)
+		return mode << 1;
+	/* else: modes 8-10 */
+	mode -= 7; /* 8 => 0b01, 9 => 0b10, 10 => 0b11 */
+	// for non-advanced, just set mode 0 if mode 8-10 was requested
+	return advanced ? (mode <<  6) : 0;
+}
+
+/* returns blink freq mode index based on bits 0-3 and 6-7 of regvals from blinking control register */
+static int regvals_to_blink_mode(int regvals, bool advanced)
+{
+	int ret, bits67;
+	if(regvals & BIT(0)) /* special case: the "always on" mode */
+		return 11;
+	/* only the advanced case (FEAT_BLINK_CTRL_ADV) has freq mode info in bits 6 and 7 */
+	bits67 = advanced ? (regvals & (BIT(6) | BIT(7))) : 0;
+	if(bits67 == 0) { /* bits 6 and 7 not set => return bits 1 to 3 */
+		ret = (regvals >> 1) & 7;
+	} else {
+		ret = bits67 >> 6;
+		ret += 7; /* 0b01 => 8 etc */
+	}
+	return ret;
+}
+
+static const u8 IT87_REG_GP_LED_CTRL_FREQ[] = {0xf9, 0xfb};
+
+const char * blink_freq_desc[] = {
+    "0.125s	OFF	0.125s ON", "0.5s OFF 0.5s ON", "2s	OFF	2s ON", "0.25s OFF 0.25s ON", "1s OFF 3s ON", "3s OFF 1s ON",
+    "2s	OFF	6s ON", "6s	OFF	2s ON", "0.5s OFF 2s ON", "1s OFF 1s ON", "4s OFF 4s ON", "ALWAYS ON"
+};
+
+static ssize_t show_gpled_blink_freq(struct device * dev, struct device_attribute *attr, char *buf)
+{
+	struct sensor_device_attribute_2 *sattr = to_sensor_dev_attr_2(attr);
+	struct it87_data *data = dev_get_drvdata(dev);
+	bool advanced = (data->features & FEAT_BLINK_CTRL_ADV) != 0;
+
+	u8 led_ctrl_reg = IT87_REG_GP_LED_CTRL_FREQ[sattr->index];
+
+	int blink_reg_val = read_gpio_reg(data, led_ctrl_reg);
+	if(blink_reg_val < 0) {
+		return blink_reg_val;
+	} else {
+		int mode = regvals_to_blink_mode(blink_reg_val, advanced);
+
+#if 0 /* useful for debugging */
+		int short_pulse = (blink_reg_val & BIT(5)) != 0;
+		int pin_map_reg_clear = (blink_reg_val & BIT(4)) != 0;
+		int blink_out_low_enab = blink_reg_val & BIT(0);
+
+		return sprintf(buf, "0x%X %d (%s) sp %d pmrc %d bole %d\n", blink_reg_val, mode,
+					   blink_freq_desc[mode], short_pulse, pin_map_reg_clear, blink_out_low_enab);
+#else
+		return sprintf(buf, "%d (%s)\n", mode, blink_freq_desc[mode]);
+#endif
+	}
+}
+
+static ssize_t set_gpled_blink_freq(struct device *dev, struct device_attribute *attr,
+                                    const char *buf, size_t count)
+{
+	struct sensor_device_attribute_2 *sattr = to_sensor_dev_attr_2(attr);
+	struct it87_data *data = dev_get_drvdata(dev);
+	u8 led_ctrl_reg = IT87_REG_GP_LED_CTRL_FREQ[sattr->index];
+	int err, blink_reg_val;
+	long val;
+	bool advanced = (data->features & FEAT_BLINK_CTRL_ADV) != 0;
+	int keep_bits = BIT(5) | BIT(4);
+	if(!advanced) /* those bits are reserved on most chips */
+		keep_bits |= (BIT(6) | BIT(7));
+
+	if (0 > kstrtol(buf, 10, &val) || val < 0 || val > 11)
+		return -EINVAL;
+	/* TODO: could support negative values for "set BIT(5) aka "Short Low Pulse Enable"
```

Comment:
Is this your TODO or someone elses?

```suggestion
	/* TODO(DanielGibson): could support negative values for "set BIT(5) aka "Short Low Pulse Enable"
```

I'd prefer to have it tagged for context. But if it was part of some other code then perhaps we can skip it.

Reply:
I mean it's all part of the LED blinking control code that is in a block here that's marked with `/* #### GP LED blinking control #### */` and `/* #### end of IT8625 LED blinking control #### */` - I think that's context enough?

I wouldn't mind changing this if you want, I don't care.

Reply:
Ah, the context I was referring to is: "ok, I found this todo here, who made it? where do I go for context?". Obviously, Git history has it, but I still like to tag TODOs with the author so that it's easy to spot, even when code moves around.